### PR TITLE
flip memory coalescing for last dim case

### PR DESCRIPTION
**Pull request by marigoold**
针对 dim = -1 时候访存无法合并的情况做了优化。
实现方式是先 flip 后写到 shared memory，然后统一从 shm 中顺序写到 global memory 中，此时可以合并访存。
对比：
```
import oneflow as flow
x = flow.randn(32, 2048).cuda()

flow._oneflow_internal.profiler.RangePush("flip_prof")
for i in range(100):
    y = flow.flip(x, [-1])
flow._oneflow_internal.profiler.RangePop()

import torch
x = torch.randn(32, 2048).cuda()

torch.cuda.nvtx.range_push("flip_prof")
for i in range(100):
    y = torch.flip(x, [-1])
torch.cuda.nvtx.range_pop()
```

- nsys 结果对比

torch kernel：
image

oneflow kernel（优化后）：
image

oneflow kernel（优化前）：
image

### CODE REVIEW

**Review by hjchen2 — APPROVED** [review]

**Review by BBuf — APPROVED** [review]

File: oneflow/user/kernels/flip_kernel.cu [hunk 1]
```diff
@@ -51,6 +52,22 @@ __global__ void FlipGpuForward(const int32_t element, const int64_t total_dims,
   }
 }
 
+template<typename T>
+__global__ void FlipLastDimGpuForward(const int32_t element, const int64_t last_dim_size,
+                                      const T* in_dptr, T* out_dptr) {
+  __shared__ T shm[kCudaThreadsNumPerBlock];
+  CUDA_1D_KERNEL_LOOP(i, element) {
```
Comment (BBuf) [inline oneflow/user/kernels/flip_kernel.cu:59]: 这个地方的坐标映射写得不直观，可以加上注释。另外可以使用pack来读写数据做延迟隐藏。另外，last_dim_size是int64_t，而i_ori是Int32_t，计算存在强制转换，可以考虑index静态派发的方式。
